extension/goplugin/util.go

```diff
@@ -74,58 +73,157 @@ func Finish(testReporter gomock.TestReporter) {
-				return nil, err
+			field.Set(reflect.ValueOf(mapValue))
+		} else if isMaybeType(kind, fieldType.Type.String()) {
+			if kind == reflect.Ptr {
+				if !mapValueExists {
+					continue
+				}
+				field.Set(reflect.New(fieldType.Type.Elem()))
+				field = field.Elem()
 			}
-			newField := reflect.New(field.Type())
-			fieldJSON := string(mapJSON)
-			fieldInterface := newField.Interface()
-			err = json.Unmarshal([]byte(fieldJSON), &fieldInterface)
-			if err != nil {
-				return nil, err
+			if err := maybeFromMap(context, propertyName, field); err != nil {
+				return err
+			}
+		} else if kind == reflect.Struct || kind == reflect.Ptr {
+			if mapValue != nil {
+				field.Set(reflect.ValueOf(reflect.New(field.Type()).Elem().Interface()))
+				if err := resourceFromMap(mapValue.(map[string]interface{}), field); err != nil {
+					return err
+				}
+			}
+		} else if kind == reflect.Slice {
+			if err := sliceToMap(context, propertyName, field); err != nil {
+				return err
 			}
-			field.Set(newField.Elem())
 		} else {
-			value := reflect.ValueOf(context[propertyName])
-			if value.IsValid() {
-				if value.Type() == field.Type() {
-					field.Set(value)
-				} else {
-					if field.Kind() == reflect.Int && value.Kind() == reflect.Float64 { // reflect treats number(N, 0) as float
-						field.SetInt(int64(value.Float()))
-					} else {
-						return nil, fmt.Errorf("invalid type of '%s' field (%s, expecting %s)", propertyName, value.Kind(), field.Kind())
+			if err := assignMapValueToField(mapValue, propertyName, field); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func assignMapValueToField(mapValue interface{}, fieldName string, field reflect.Value) error {
+	value := reflect.ValueOf(mapValue)
+	if value.IsValid() {
+		if value.Type() == field.Type() {
+			field.Set(value)
+		} else {
+			if field.Kind() == reflect.Int && value.Kind() == reflect.Float64 { // reflect treats number(N, 0) as float
+				field.SetInt(int64(value.Float()))
+			} else {
+				return fmt.Errorf("invalid type of '%s' field (%s, expecting %s)", fieldName, value.Kind(), field.Kind())
+			}
+		}
+	}
+	return nil
+}
+
+func isMaybeType(kind reflect.Kind, typeName string) bool {
+	return anyOfKinds(kind, []reflect.Kind{reflect.Ptr, reflect.Struct}) && strings.HasPrefix(typeName, "goext.Maybe")
+}
+
+func maybeFromMap(context map[string]interface{}, fieldName string, field reflect.Value) error {
+	if mapValue, ok := context[fieldName]; !ok {
+		field.FieldByName("State").SetInt(int64(goext.MaybeUndefined))
+	} else if mapValue == nil {
+		field.FieldByName("State").SetInt(int64(goext.MaybeNull))
+	} else {
+		field.FieldByName("State").SetInt(int64(goext.MaybeNotNull))
+		if err := assignMapValueToField(mapValue, fieldName, field.FieldByName("Value")); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func sliceToMap(context map[string]interface{}, fieldName string, field reflect.Value) error {
+	if v, ok := context[fieldName]; ok {
```

Comment:
It better to implement from specific cases to generic cases: first handle any specific cases and only after interpret the generic case. This way there will be less indentation.